### PR TITLE
Refactoring of all functional tests under `functest` dir

### PR DESCRIPTION
**Pull request by aruniiird**
Signed-off-by: Arun Kumar Mohan <amohan@redhat.com>

### CODE REVIEW

**Comment by aruniiird** [comment]
@umangachapagain , @jarrpa 
Need an `ok-to-test` kick start the CI checks

**Comment by aruniiird** [comment]
@jarrpa , I have divided the commits into several logical ones. Please take a look...

File: functests/storageclusterinitialization_test.go [hunk 1]
```diff
@@ -483,6 +512,44 @@ func (scExpctObj *storageClassExpect) ExpectReconcile() error {
...
+	ExpectDelete(bool)
+	ExpectReconcile() error
+}
+
+func getAllValidExpectResources(scInitObj *SCInit, tobeDeleted bool) []ExpectTypeI {
+	var expctRs = []ExpectTypeI{
+		scInitObj.createCephFilesystemExpect(tobeDeleted),
+		scInitObj.createStorageClassExpect(tobeDeleted),
+	}
+	// CephBlockPool can only be added if it is not tobe deleted (ie; 'tobeDeleted' flag is false)
```
Comment (jarrpa) [inline functests/storageclusterinitialization_test.go:529]:
```suggestion
	// CephBlockPool can only be added if it is not to be deleted (ie; 'tobeDeleted' flag is false)
```

File: functests/storageclusterinitialization_test.go [hunk 2]
```diff
@@ -525,123 +592,43 @@ func StorageClusterInitializationTest() {
...
-					By("Modifying CephObjectStore (expected tobe deleted)")
-					err = cephObjStoreExpect.ExpectModify()
-					Expect(err).To(BeNil())
-
-					cephObjStoreUserExpect = scInitObj.createCephObjectStoreUserExpect(true)
-					By("Modifying CephObjectStoreUser (expected tobe deleted)")
-					err = cephObjStoreUserExpect.ExpectModify()
+				var expectResources = getAllValidExpectResources(scInitObj, true)
+				for _, eachResource := range expectResources {
+					By(fmt.Sprintf("Modifying %s (expected tobe deleted)",
```
Comment (jarrpa) [inline functests/storageclusterinitialization_test.go:617]:
```suggestion
					By(fmt.Sprintf("Modifying %s (expected to be deleted)",
```

**Comment by aruniiird** [comment]
I have addressed all the review comments except this one,

> For your last two commits, is it possible to simply have one struct Expect{} type that takes a kind string and an obj runtime.Object?

I am unable to generalize the functions `ExpectReconcile()` and `ExpectModify()` , with the `runtime.Object` interface, as these functions require specific / concrete object's `Spec` values (like `Gateway` or `DataPools[]` or `Store` etc). Example lines,

https://github.com/openshift/ocs-operator/blob/95db65455429b7c93f399cc3e41f4d6c29d9d2b7/functests/storageclusterinitialization_test.go#L361
https://github.com/openshift/ocs-operator/blob/95db65455429b7c93f399cc3e41f4d6c29d9d2b7/functests/storageclusterinitialization_test.go#L287
https://github.com/openshift/ocs-operator/blob/95db65455429b7c93f399cc3e41f4d6c29d9d2b7/functests/storageclusterinitialization_test.go#L213

Solutions which I come up with to generalize all objects (bringing another common interfaces to get the specific `Spec`'s value) are making the tests more and more complicated.

**Pull request merged**